tools/gfx/debug-layer.cpp

```diff
@@ -0,0 +1,948 @@
+    return baseObject->resize(width, height);
+}
+
+void DebugSwapchain::maybeRebuildImageList()
+{
+    SLANG_GFX_API_FUNC;
+    if (m_images.getCount() != 0)
+        return;
+    m_images.clearAndDeallocate();
+    for (uint32_t i = 0; i < baseObject->getDesc().imageCount; i++)
+    {
+        RefPtr<DebugTextureResource> image = new DebugTextureResource();
+        baseObject->getImage(i, image->baseObject.writeRef());
+        m_images.add(image);
+    }
+}
+
+slang::TypeLayoutReflection* DebugShaderObject::getElementTypeLayout()
+{
+    SLANG_GFX_API_FUNC;
+    return baseObject->getElementTypeLayout();
+}
+
+UInt DebugShaderObject::getEntryPointCount()
+{
+    SLANG_GFX_API_FUNC;
+    return baseObject->getEntryPointCount();
+}
+
+Result DebugShaderObject::getEntryPoint(UInt index, IShaderObject** entryPoint)
+{
+    SLANG_GFX_API_FUNC;
+    if (m_entryPoints.getCount() == 0)
+    {
+        for (UInt i = 0; i < getEntryPointCount(); i++)
+        {
+            RefPtr<DebugShaderObject> entryPointObj = new DebugShaderObject();
+            SLANG_RETURN_ON_FAIL(
+                baseObject->getEntryPoint(i, entryPointObj->baseObject.writeRef()));
+            m_entryPoints.add(entryPointObj);
+        }
+    }
+    if (index > (UInt)m_entryPoints.getCount())
+    {
+        GFX_DIAGNOSE_ERROR("`index` must not exceed `entryPointCount`.");
+        return SLANG_FAIL;
+    }
+    returnComPtr(entryPoint, m_entryPoints[index]);
+    return SLANG_OK;
+}
+
+Result DebugShaderObject::setData(ShaderOffset const& offset, void const* data, size_t size)
+{
+    SLANG_GFX_API_FUNC;
+    return baseObject->setData(offset, data, size);
+}
+
+Result DebugShaderObject::getObject(ShaderOffset const& offset, IShaderObject** object)
+{
+    SLANG_GFX_API_FUNC;
+
+    ComPtr<IShaderObject> innerObject;
+    auto resultCode = baseObject->getObject(offset, innerObject.writeRef());
+    SLANG_RETURN_ON_FAIL(resultCode);
+    RefPtr<DebugShaderObject> debugShaderObject;
+    if (m_objects.TryGetValue(ShaderOffsetKey{offset}, debugShaderObject))
+    {
+        if (debugShaderObject->baseObject == innerObject)
+        {
+            returnComPtr(object, debugShaderObject);
+            return resultCode;
+        }
+    }
+    debugShaderObject = new DebugShaderObject();
+    debugShaderObject->baseObject = innerObject;
+    m_objects[ShaderOffsetKey{offset}] = debugShaderObject;
+    returnComPtr(object, debugShaderObject);
+    return resultCode;
+}
+
+Result DebugShaderObject::setObject(ShaderOffset const& offset, IShaderObject* object)
+{
+    SLANG_GFX_API_FUNC;
+    auto objectImpl = static_cast<DebugShaderObject*>(object);
+    m_objects[ShaderOffsetKey{offset}] = objectImpl;
+    return baseObject->setObject(offset, objectImpl->baseObject.get());
+}
+
+Result DebugShaderObject::setResource(ShaderOffset const& offset, IResourceView* resourceView)
+{
+    SLANG_GFX_API_FUNC;
+    auto viewImpl = static_cast<DebugResourceView*>(resourceView);
```

Comment:
Yeah, we really want operations like this to be able to do validation on the `offset` being passed in and make sure that it properly identifies a binding range of an appropriate type for the review being passed in. We'd want to validate that the resource was allocated to support whatever state/usage the range requires.

It would also seem important to validate at draw/dispatch time that the objects bound to the pipeline via shader objects are in the required state for the operations being done.